jena-arq/src/main/java/org/apache/jena/sparql/engine/iterator/QueryIterLateral.java

```diff
@@ -292,17 +295,33 @@ public Op transform(OpTable opTable) {
             // By the assignment restriction, the binding only needs to be added to each row of the table.
             Table table = opTable.getTable();
             // Table vars.
-            List<Var> vars = new ArrayList<>(table.getVars());
-            binding.vars().forEachRemaining(vars::add);
-            TableN table2 = new TableN(vars);
+            List<Var> tableVars = table.getVars();
+            List<Var> vars = new ArrayList<>(tableVars);
+
+            // Track variables that appear both in the table and the binding.
+            List<Var> commonVars = new ArrayList<>();
+
+            // Index variables in a set if there are more than a few of them.
+            Collection<Var> tableVarsIndex = tableVars.size() > 4 ? new HashSet<>(tableVars) : tableVars;
+            binding.vars().forEachRemaining(v -> {
+                if (tableVarsIndex.contains(v)) {
```

Comment:
The table vars can't contain a binding var. That case is caught by the 

https://github.com/w3c/sparql-dev/blob/main/SEP/SEP-0007/sep-0007.md  
"Disallow syntactic forms that set variables that may already be present in the current row."

It's tested in `SyntaxVarScope.checkLATEAL`.

(I removed the `if`, left the `else` and the second test passes). 

If so, it simplifies to `vars.add(v);`
And hence "vars = tableVars" (copy not needed).